hudi-spark-datasource/hudi-spark2/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/Spark2HoodieParquetFileFormat.scala

```diff
@@ -0,0 +1,296 @@
+
+      ParquetWriteSupport.setSchema(requiredSchema, hadoopConf)
+
+      // Sets flags for `ParquetToSparkSchemaConverter`
+      hadoopConf.setBoolean(
+        SQLConf.PARQUET_BINARY_AS_STRING.key,
+        sparkSession.sessionState.conf.isParquetBinaryAsString)
+      hadoopConf.setBoolean(
+        SQLConf.PARQUET_INT96_AS_TIMESTAMP.key,
+        sparkSession.sessionState.conf.isParquetINT96AsTimestamp)
+
+      val broadcastedHadoopConf =
+        sparkSession.sparkContext.broadcast(new SerializableConfiguration(hadoopConf))
+
+      // TODO: if you move this into the closure it reverts to the default values.
+      // If true, enable using the custom RecordReader for parquet. This only works for
+      // a subset of the types (no complex types).
+      val resultSchema = StructType(partitionSchema.fields ++ requiredSchema.fields)
+      val sqlConf = sparkSession.sessionState.conf
+      val enableOffHeapColumnVector = sqlConf.offHeapColumnVectorEnabled
+      val enableVectorizedReader: Boolean =
+        sqlConf.parquetVectorizedReaderEnabled &&
+          resultSchema.forall(_.dataType.isInstanceOf[AtomicType])
+      val enableRecordFilter: Boolean = sqlConf.parquetRecordFilterEnabled
+      val timestampConversion: Boolean = sqlConf.isParquetINT96TimestampConversion
+      val capacity = sqlConf.parquetVectorizedReaderBatchSize
+      val enableParquetFilterPushDown: Boolean = sqlConf.parquetFilterPushDown
+      // Whole stage codegen (PhysicalRDD) is able to deal with batches directly
+      val returningBatch = supportBatch(sparkSession, resultSchema)
+      val pushDownDate = sqlConf.parquetFilterPushDownDate
+      val pushDownTimestamp = sqlConf.parquetFilterPushDownTimestamp
+      val pushDownDecimal = sqlConf.parquetFilterPushDownDecimal
+      val pushDownStringStartWith = sqlConf.parquetFilterPushDownStringStartWith
+      val pushDownInFilterThreshold = sqlConf.parquetFilterPushDownInFilterThreshold
+      val isCaseSensitive = sqlConf.caseSensitiveAnalysis
+
+      (file: PartitionedFile) => {
+        assert(file.partitionValues.numFields == partitionSchema.size)
+
+        val fileSplit =
+          new FileSplit(new Path(new URI(file.filePath)), file.start, file.length, Array.empty)
+        val filePath = fileSplit.getPath
+
+        val split =
+          new org.apache.parquet.hadoop.ParquetInputSplit(
+            filePath,
+            fileSplit.getStart,
+            fileSplit.getStart + fileSplit.getLength,
+            fileSplit.getLength,
+            fileSplit.getLocations,
+            null)
+
+        val sharedConf = broadcastedHadoopConf.value.value
+
+        // do deal with internalSchema
+        val internalSchemaString = sharedConf.get(SparkSchemaUtils.HOODIE_QUERY_SCHEMA)
+        val internalSchemaChangeEnabled = if (internalSchemaString.isEmpty) false else true
+        // querySchema must be a pruned schema.
+        val querySchema = InternalSchemaParser.fromJson(internalSchemaString)
```

Comment:
Corner case: If `internalSchemaString` is empty then `querySchema` could be null. Let's handle this.

Reply:
internalSchemaChangeEnabled will deal with this corner case， rebuildFilterFromParquet will also deal with this corner case， donot worried